### ISSUE TITLE
Different padding boundaries for different directions. 

### ISSUE THREAD

**Issue by tanmayshankar** [tensorflow/tensorflow#3247]
Hello, 

I recently started to use TensorFlow to implement and idea using 3D convolutions. It requires me to use different padding boundary conditions along spatial and temporal directions (dimensions) of the tensor (I need `padding='SAME'` for spatial and `padding='VALID'` for temporal directions).  `tf.nn.conv3d` and `tf.nn.max_pool3d` specify padding for the operation as a whole. 

Is there any way to specify padding type for each dimension individually, like how the stride is specified? I don't mind contributing for this if necessary, if someone could direct me on how to add this feature or create a new op for it.

**Comment by vrv** [comment]
Unfortunately at the moment there is no way to specify different padding algorithms for different dimensions.

The ideal way to handle this (and other future feature requests like this) would be to support explicit padding -- the user passes the amount of padding on the 'top' and 'bottom' of each dimension, so they have full control of the algorithm.

To do this, code that calls into eigen_spatial_convolution.h and cuboid_convolution.h (etc.) would need to call a version of the interfaces that passes the explicit dimensions.  Right now we pass 'padding algorithms.'   For example: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/eigen_spatial_convolutions.h#L932 would need to be changed to take explicit padding values instead of 'padding_type'.  We could then change the API that has padding_type to call into the one that provides the explicit padding dimensions, to maximize code reuse.

This would be a major/difficult change for someone not familiar with the codebase, and it's something we can track internally.

**Comment by jmchen-g** [comment]
Tracked this internally so I will close this one here.

**Comment by tanmayshankar** [comment]
Hi Vijay and Jianmin, thanks for taking the time to look into this. 

I agree that controlling padding amounts of each dimension is probably the most general scenario, but  I also like the existing abstraction of padding types in place. 

Until this feature is brought in, would it be possible for me to define an alternate operation to conv3d, 
where the amount of padding for planes, rows and columns is decided case-wise based on the padding type? 
It seems that the conv_ops_3d.cc seems to be doing something similar (If I'm right, the VALID padding is some sort of default case, and SAME padding is built over that?), except that it clubs rows, columns and planes within those if statements checking padding types.

**Comment by Nimitz14** [comment]
Has this ever been looked in to?

**Comment by florianletsch** [comment]
Here's the solution I've used: Set padding to be 'VALID' (= no padding), but pad manually in the spatial dimension using `tf.pad`:

```py
# pad spatially
paddings = tf.constant([[0, 0], [0, 0], [1, 1], [1, 1], [0, 0]])
padded = tf.pad(inputs, paddings)

# convolution without automatic padding
tf.nn.conv3d(inputs, filters, strides=[1, 1, 1, 1, 1], padding='VALID)
```
